src/con_duct/cli.py

```diff
@@ -0,0 +1,474 @@
+
+environment variables:
+  Many duct options can be configured by environment variables (which are
+  overridden by command line options).
+
+  DUCT_LOG_LEVEL: see --log-level
+  DUCT_OUTPUT_PREFIX: see --output-prefix
+  DUCT_SUMMARY_FORMAT: see --summary-format
+  DUCT_SAMPLE_INTERVAL: see --sample-interval
+  DUCT_REPORT_INTERVAL: see --report-interval
+  DUCT_CAPTURE_OUTPUTS: see --capture-outputs
+  DUCT_MESSAGE: see --message
+"""
+
+
+class CustomHelpFormatter(argparse.ArgumentDefaultsHelpFormatter):
+    def _fill_text(self, text: str, width: int, _indent: str) -> str:
+        return "\n".join([textwrap.fill(line, width) for line in text.splitlines()])
+
+
+def create_common_parser() -> argparse.ArgumentParser:
+    """Create a parser with common arguments shared across all commands."""
+    parser = argparse.ArgumentParser(add_help=False)
+    parser.add_argument(
+        "-l",
+        "--log-level",
+        default=DEFAULT_LOG_LEVEL,
+        choices=("NONE", "CRITICAL", "ERROR", "WARNING", "INFO", "DEBUG"),
+        type=str.upper,
+        help="Level of log output to stderr, use NONE to entirely disable.",
+    )
+    parser.add_argument(
+        "-q",
+        "--quiet",
+        action="store_true",
+        help="[deprecated, use log level NONE] Disable duct logging output (to stderr)",
+    )
+    return parser
+
+
+def setup_logging(args: argparse.Namespace) -> None:
+    """Configure logging based on parsed arguments.
+
+    Handles both --log-level and --quiet flags, applying them consistently
+    across all subcommands.
+    """
+    log_level = args.log_level
+
+    # Handle deprecated --quiet flag
+    if args.quiet:
+        log_level = "NONE"
+
+    # Special case: NONE means disable all logging
+    if log_level == "NONE":
+        logging.disable(logging.CRITICAL)
+    else:
+        logging.basicConfig(
+            format="%(asctime)s [%(levelname)-8s] %(name)s: %(message)s",
+            datefmt="%Y-%m-%dT%H:%M:%S%z",
+            level=log_level,
+        )
+
+
+@dataclass
+class RunArguments:
+    command: str
+    command_args: list[str]
+    output_prefix: str
+    sample_interval: float
+    report_interval: float
+    fail_time: float
+    clobber: bool
+    capture_outputs: Outputs
+    outputs: Outputs
+    record_types: RecordTypes
+    summary_format: str
+    colors: bool
+    log_level: str
+    quiet: bool
+    session_mode: SessionMode
+    message: str = ""
+
+    def __post_init__(self) -> None:
+        if self.report_interval < self.sample_interval:
+            raise argparse.ArgumentError(
+                None,
+                "--report-interval must be greater than or equal to --sample-interval.",
+            )
+
+    def execute(self) -> int:
+        """Execute the command with these arguments.
+
+        This is a convenience method for tests and library usage.
+        """
+        return duct_execute(
+            command=self.command,
```

Comment:
is there 1-to-1 correspondence between `RunArguments` and ArgumentParser's collection of attributes?

ATM it feels very heavy in how much of `X=Y.X` there are -- refactor/get helper to just use a loop with `getattr` here, in `from_argv` and `run_command` .

Reply:
after removing RunArguments this breaks down to just removing extra args and passing kwargs